file_writer_control/CommandChannel.py

```diff
@@ -0,0 +1,217 @@
+        self.map_of_commands: Dict[str, CommandStatus] = {}
+        self.run_thread = True
+        self.thread = threading.Thread(
+            target=thread_function, daemon=True, kwargs=thread_kwargs
+        )
+        self.thread.start()
+
+        def do_exit():
+            self.stop_thread()
+
+        atexit.register(do_exit)
+
+    def add_job_id(self, job_id: str):
+        """
+        Add a job identifier to the list of known jobs before it has been encountered on the command topic.
+        :param job_id: The identifier of the new job.
+        """
+        if job_id not in self.map_of_jobs:
+            self.map_of_jobs[job_id] = JobStatus(job_id)
+
+    def add_command_id(self, job_id: str, command_id: str):
+        """
+        Add a command identifier to the list of known commands before it has been encountered on the command topic.
+        :param job_id: The job identifier of the new command.
+        :param command_id: The identifier of the new command.
+        """
+        if command_id not in self.map_of_commands:
+            self.map_of_commands[command_id] = CommandStatus(job_id, command_id)
+
+    def stop_thread(self):
+        """
+        Stop the thread that is continuously getting command topic messages in the background. Should only be called if
+        we are about to get rid of the current instance of CommandChannel.
+        """
+        self.to_thread_queue.put("exit")
+        try:
+            self.thread.join()
+        except RuntimeError:
+            pass
+
+    def __del__(self):
+        self.stop_thread()
+
+    def update_workers(self):
+        """
+        Update the list of known workers, jobs and commands. This is a non-blocking call but it might take some time
+        to execute if the queue of updates is long. This member function is called by many of the other member functions
+        in this class.
+        """
+        while not self.status_queue.empty():
+            status_update = self.status_queue.get()
+            if type(status_update) is WorkerStatus:
+                if status_update.service_id not in self.map_of_workers:
+                    self.map_of_workers[status_update.service_id] = status_update
+                self.map_of_workers[status_update.service_id].update_status(
+                    status_update
+                )
+            elif type(status_update) is JobStatus:
+                if status_update.job_id not in self.map_of_jobs:
+                    self.map_of_jobs[status_update.job_id] = status_update
+                self.map_of_jobs[status_update.job_id].update_status(status_update)
+            elif type(status_update) is CommandStatus:
+                if status_update.command_id not in self.map_of_commands:
+                    self.map_of_commands[status_update.command_id] = status_update
+                self.map_of_commands[status_update.command_id].update_status(
+                    status_update
```

Comment:
To be honest I do not like this much of nested logic flows (they are highly problematic to maintain and very bug prone). I do not know if they are escapable here but there should be a more pythonic way to handle this. If not, no worries.

Reply:
I partially agree with you. I have re-written some of the code.